BUILD

```diff
@@ -0,0 +1,393 @@
+    hdrs = ["include/benchmark/benchmark.h"],
+    strip_include_prefix = "include",
+ )
+
+cc_library(
+    name = "benchmark_register",
+    srcs = ["src/benchmark_register.cc"],
+    deps = [
+        ":benchmark_api_internal",
+        ":benchmark_header",
+        ":check",
+        ":commandlineflags",
+        ":complexity",
+        ":internal_macros",
+        ":mutex",
+        ":re",
+        ":statistics",
+        ":string_util",
+        ":timers",
+    ],
+ )
+
+cc_library(
+    name = "check",
+    hdrs = ["src/check.h"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+        ":internal_macros",
+        ":log",
+    ],
+)
+
+cc_library(
+    name = "colorprint",
+    hdrs = ["src/colorprint.h"],
+    srcs = ["src/colorprint.cc"],
+    strip_include_prefix = "src",
+    deps = [
+        ":check",
+        ":internal_macros",
+    ]
+)
+
+cc_library(
+    name = "commandlineflags",
+    hdrs = ["src/commandlineflags.h"],
+    srcs = ["src/commandlineflags.cc"],
+    strip_include_prefix = "src",
+    deps = [
+    ]
+)
+
+cc_library(
+    name = "complexity",
+    hdrs = ["src/complexity.h"],
+    srcs = ["src/complexity.cc"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+        ":check",
+    ]
+)
+
+cc_library(
+    name = "console_reporter",
+    srcs = ["src/console_reporter.cc"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+        ":commandlineflags",
+        ":complexity",
+        ":colorprint",
+        ":counter",
+        ":string_util",
+        ":timers",
+    ]
+)
+
+cc_library(
+    name = "counter",
+    hdrs = ["src/counter.h"],
+    srcs = ["src/counter.cc"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+    ]
+)
+
+cc_library(
+    name = "csv_reporter",
+    srcs = ["src/csv_reporter.cc"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+        ":complexity",
+        ":string_util",
+        ":timers",
+    ]
+)
+
+cc_library(
+    name = "cycleclock",
+    hdrs = ["src/cycleclock.h"],
+    strip_include_prefix = "src",
+    deps = [
+    ]
+)
+cc_library(
+    name = "internal_macros",
+    hdrs = ["src/internal_macros.h"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+    ],
+)
+
+cc_library(
+    name = "json_reporter",
+    srcs = ["src/json_reporter.cc"],
+    strip_include_prefix = "src",
+    deps = [
+        ":benchmark_header",
+        ":complexity",
+        ":string_util",
+        ":timers",
+    ]
+)
+
+cc_library(
+    name = "log",
+    hdrs = ["src/log.h"],
+    strip_include_prefix = "src",
+)
+
+cc_library(
+    name = "mutex",
+    hdrs = ["src/mutex.h"],
+    strip_include_prefix = "src",
+    deps = [
+    ]
+)
+
+cc_library(
+    name = "re",
+    hdrs = ["src/re.h"],
+    strip_include_prefix = "src",
+    deps = [
+    ],
+    defines = ["HAVE_POSIX_REGEX"],
```

Comment:
note, this won't work for everyone. this is one of the reasons i held off; no way to do this kind of dynamic configuration.

Reply:
It's not perfect agreed, but it's also not unprecedented. gflags has a similar hardcoding (https://github.com/gflags/gflags/blob/master/bazel/gflags.bzl). I started a patch for moving some of the defines there to run off of a config but haven't gotten to a PR for that yet (googletest does something similar for linkopts: https://github.com/google/googletest/blob/master/BUILD.bazel).

But I know I don't have complete access to all the build modes I might need. Is there an automated build tool I can use to ensure coverage?

Reply:
the travis and appveyor builds are probably the closest we have for covering the various platforms/compilers. there are other platforms that people run on, but we don't have a way to cover every combination.